django/db/models/sql/query.py

```diff
@@ -1401,9 +1402,42 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
         key field for example).
         """
         joins = [alias]
-        # First, generate the path for the names
-        path, final_field, targets, rest = self.names_to_path(
-            names, opts, allow_many, fail_on_missing=True)
+        transforms = []
+        # Try resolving all the names as fields first, but if there is an error
+        # progressively treat trailing names as lookups until we can resolve
+        # a field
+        last_field_exception = None
+        for pivot in range(len(names), 0, -1):
+            try:
+                path, final_field, targets, rest = self.names_to_path(
+                    names[:pivot], opts, allow_many, fail_on_missing=True)
+            except (FieldError) as exc:
+                if pivot == 1:
+                    # The first item cannot be a lookup, so it's safe
+                    # to raise the field error here.
+                    raise
+                else:
+                    last_field_exception = exc
+            else:
+                # The transforms are the remaining items that couldn't be
+                # resolved into fields
+                transforms = names[pivot:]
+                break
+        for name in transforms:
+            try:
+                if hasattr(final_field, 'resolve_expression'):
+                    final_field = self.try_transform(final_field, name)
+                else:
+                    final_field = self.try_transform(final_field.get_col(alias), name)
```

Comment:
Your suggestion here is actually wrong, it needs to be in a temporary variable, not final_field, otherwise if an exception is raised final_field is left as a Col rather than a Field.